yksgi/src/lib.rs

```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+        // Keep track of the addresses of the current and previous frame so we can update the RBP
+        // register as needed.
+        let mut currframe = btmframeaddr;
+        let mut nextframe = btmframeaddr;
+
+        for (i, frame, rec, pinfo) in smaps {
+            debug_assert!(rec.size != u64::MAX);
+
+            // WRITE RBP
+            // If the current frame has pushed RBP we need to do the same (unless we are processing
+            // the bottom-most frame).
+            if pinfo.hasfp && i > 0 {
+                rsp = unsafe { rsp.offset(-8) };
+                rbp = rsp;
+                unsafe { ptr::write(rsp as *mut u64, currframe as u64) };
+            }
+
+            // Now that we've (potentially) written the last frame's address, update currframe to
+            // the actual current frame.
+            currframe = nextframe;
+
+            // Update RBP to represent this frame's address.
+            if pinfo.hasfp {
+                registers[6] = currframe as u64;
+            }
+
+            // Calculate the next frame's address by substracting its size (plus return address)
+            // from the current frame's address.
+            nextframe = unsafe { currframe.offset(-isize::try_from(rec.size).unwrap() - 8) };
+
+            // WRITE CALLEE-SAVED REGISTERS
+            // Now we write any callee-saved registers onto the new stack. Note, that if we have
+            // pushed RBP above (which includes adjusting RBP) we need to temporarily re-adjust our
+            // pointer. This is because the CSR index calculates from the bottom of the frame, not
+            // from RBP. For example, a typical prologue looks like this:
+            //   push rbp
+            //   mov rbp, rsp
+            //   push rbx     # this has index -2
+            //   push r14     # this has index -3
+            if i > 0 {
+                for (reg, idx) in &pinfo.csrs {
+                    let mut tmp = unsafe { rbp.offset(isize::try_from(idx * 8).unwrap()) };
+                    if pinfo.hasfp {
+                        tmp = unsafe { tmp.offset(8) };
+                    }
+                    let val = registers[usize::from(*reg)];
+                    unsafe { ptr::write(tmp as *mut u64, val) };
+                }
+            }
+
+            // WRITE STACKMAP LOCATIONS.
+            // Now write all live variables to the new stack in the order they are listed in the
+            // AOT stackmap call.
+            let smcall = unsafe { get_stackmap_call(frame.pc) };
+            for (j, lv) in rec.live_vars.iter().enumerate() {
+                // Adjust the operand index by 2 to skip stackmap ID and shadow bytes.
+                let op = smcall.get_operand(u32::try_from(j+2).unwrap());
+                let val = if let Some(val) = frame.get(&op) {
+                    val.val
+                } else {
+                    todo!()
+                    //println!("probably return pointer which will disappear");
```

Comment:
Commented code (you can put text in `todo` if you want e.g. `todo!("probably return a pointer which will disappear")`).

Reply:
Actually, I need to reinvestigate this code here. I think this can be removed entirely now.

Reply:
[bc9be2b](https://github.com/ykjit/yk/pull/578/commits/bc9be2bb1a42334c625104ce5309e9a7b5f0cb8b)